log/alt_experimental_level/benchmark_test.go

```diff
@@ -0,0 +1,85 @@
+package level_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/alt_experimental_level"
+)
+
+func BenchmarkNopBaseline(b *testing.B) {
+	singleRecordBenchmarkRunner(b, log.NewNopLogger())
+}
+
+func BenchmarkNopDisallowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingInfoAndAbove(log.NewNopLogger()))
+}
+
+func BenchmarkNopAllowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingAll(log.NewNopLogger()))
+}
+
+func BenchmarkJSONBaseline(b *testing.B) {
+	singleRecordBenchmarkRunner(b, log.NewJSONLogger(ioutil.Discard))
+}
+
+func BenchmarkJSONDisallowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingInfoAndAbove(log.NewJSONLogger(ioutil.Discard)))
+}
+
+func BenchmarkJSONAllowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingAll(log.NewJSONLogger(ioutil.Discard)))
+}
+
+func BenchmarkLogfmtBaseline(b *testing.B) {
+	singleRecordBenchmarkRunner(b, log.NewLogfmtLogger(ioutil.Discard))
+}
+
+func BenchmarkLogfmtDisallowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingInfoAndAbove(log.NewLogfmtLogger(ioutil.Discard)))
+}
+
+func BenchmarkLogfmtAllowedLevel(b *testing.B) {
+	singleRecordBenchmarkRunner(b,
+		level.AllowingAll(log.NewLogfmtLogger(ioutil.Discard)))
+}
+
+func singleRecordBenchmarkRunner(b *testing.B, logger log.Logger) {
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		level.Debug(logger).Log("foo", "bar")
+	}
+}
+
+func BenchmarkDroppedRecords(b *testing.B) {
+	logger := log.NewNopLogger()
+	logger = log.NewContext(logger).With("ts", log.DefaultTimestamp, "caller", log.DefaultCaller)
+	for _, dropped := range []uint{1, 3, 9, 99, 999} {
+		b.Run(fmt.Sprintf("%d-of-%d", dropped, dropped+1), func(b *testing.B) {
+			manyRecordBenchmarkRunner(b, logger, dropped)
+		})
+	}
+}
+
+func manyRecordBenchmarkRunner(b *testing.B, logger log.Logger, droppedRecords uint) {
+	logger = level.AllowingInfoAndAbove(logger)
+	debug := level.Debug(logger)
+	info := level.Info(logger)
+	b.ResetTimer()
+	b.ReportAllocs()
+	for i := 0; i < b.N; i++ {
+		// Only this one will be retained.
+		info.Log("foo", "bar")
+		for ; droppedRecords != 0; droppedRecords-- {
+			debug.Log("baz", "quux")
+		}
+	}
+}
```

Comment:
I'm sorry again that it took me this long to respond.

You are right that I introduced a defect into the benchmark function when I moved some statements around, mistakenly mutating the function parameter _inside_ the benchmark loop. When I first wrote it, that loop that decays the "droppedRecords" parameter was inside a function called on from the benchmark loop.

In any case, your version does something different from what I had intended. Your drops a record when `testing.B.N` has a certain value. Instead, I intended to drop the "droppedRecords" parameters value and only emit one, so if we say we're benchmarking 99 dropped records, our test involves emitting one record and dropping 99, and doing so on each iteration of the benchmark.

Commit 27c16543e81b6b3481999c76fc9dc6ac2ade8453 restores the behavior I intended. However, the measurements aren't much different from yours:
```
name                          old time/op    new time/op    delta
DroppedRecords/1-of-2-8         2.01µs ± 0%    1.72µs ± 0%  -14.36%
DroppedRecords/3-of-4-8         2.56µs ± 0%    1.95µs ± 0%  -23.97%
DroppedRecords/9-of-10-8        4.14µs ± 0%    2.67µs ± 0%  -35.42%
DroppedRecords/99-of-100-8      27.6µs ± 0%    13.0µs ± 0%  -53.05%
DroppedRecords/999-of-1000-8     268µs ± 0%     116µs ± 0%  -56.65%

name                          old alloc/op   new alloc/op   delta
DroppedRecords/1-of-2-8           463B ± 0%      399B ± 0%  -13.82%
DroppedRecords/3-of-4-8           719B ± 0%      527B ± 0%  -26.70%
DroppedRecords/9-of-10-8        1.49kB ± 0%    0.91kB ± 0%  -38.74%
DroppedRecords/99-of-100-8      13.0kB ± 0%     6.7kB ± 0%  -48.71%
DroppedRecords/999-of-1000-8     128kB ± 0%      64kB ± 0%  -49.87%

name                          old allocs/op  new allocs/op  delta
DroppedRecords/1-of-2-8           12.0 ± 0%      10.0 ± 0%  -16.67%
DroppedRecords/3-of-4-8           20.0 ± 0%      16.0 ± 0%  -20.00%
DroppedRecords/9-of-10-8          44.0 ± 0%      34.0 ± 0%  -22.73%
DroppedRecords/99-of-100-8         404 ± 0%       304 ± 0%  -24.75%
DroppedRecords/999-of-1000-8     4.00k ± 0%     3.00k ± 0%  -24.98%
```